### DISCUSSION TITLE
Proposal 13: upgrade onchain move language version to v6

### DISCUSSION THREAD

**Discussion by nkysg** [starcoinorg/starcoin#3852] [DaoProposal]
## English guide
Voters need to update to the latest version  [v1.13.0](https://github.com/starcoinorg/starcoin/releases/tag/v1.13.0).
This proposal will upgrade the on-chain  configuration of Move bytecode version to v6.

## Upgrade Proposal Information

- Proposal Transaction: https://stcscan.io/main/transactions/detail/0x786d69ec448b787690244df0adb7051d5990e8db32f06f5c555cad5fb5b7d4f5
- Proposer: [0xed9ea1f3533c14e1b52d9ff6475776ba](https://stcscan.io/main/address/0xed9ea1f3533c14e1b52d9ff6475776ba)
- Proposal id: 13

## Upgrade process

1. 7 days of voting will be held from the launch of the upgrade proposal today.
2. After 7 days, if the proposal is approved, the proposal will go into a public period of 1 day. If it does not pass, the proposal fails.
3. after the public period, anyone can initiate a transaction to write the new configuration to the chain. After this upgrade transaction is executed, the Move new feature is activated. **Nodes that any node who have not upgraded to starcoin v1.13.0 may fork from the main network at that time, as the new feature of Move includes a hard fork feature. **

For more information on Starcoin governance, please see:

* [Governance section of the Starcoin technical white paper](https://developer.starcoin.org/zh/sips/sip-2/)
* [Lifecycle of on-chain governance](https://developer.starcoin.org/zh/key_concepts/dao_governance/)
* [Modifying DAO settings through the governance mechanism](https://developer.starcoin.org/zh/cli/modify_dao_config/)

Note: After voting, the STC of the current account will be pledged to the contract until the votes ends, and the current voting period is 7 days. Active refunds during this period will be considered as abandoned votes. ** To encourage participation in on-chain governance, incentives for on-chain governance will be introduced in the future. Users who participate in voting this time will be compensated according to the incentive strategy. **

### Preparation

1. Download latest starcoin version and synchronize the blocks on the main network, then enter the starcoin console.
2. Import your account into the node's built-in wallet or withdraw coins from the exchange to the node's wallet address.

### Operation commands

1. Check the voting status

```
dev call --function 0x1::Dao::proposal_state -t 0x1::STC::STC -t 0x1::OnChainConfigDao::OnChainConfigUpdate<0x1::LanguageVersion:: LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba --arg 13
```
The first argument is the address of the proposer, the second argument is the proposal id, and the returned status is 1 for waiting status, 2 for voting, and 4 for passing the ballot.

2. Execute voting

Voting is performed by the following command.

```
account execute-function -b -s your_address --function 0x1::DaoVoteScripts::cast_vote -t 0x1::STC::STC -t 0x1::OnChainConfigDao:: OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba --arg 13 --arg true --arg 1000000000u128
```

The first argument is the address of the proposer, the second argument is the proposal id, the third argument indicates agree and disagree via true/false, and the fourth argument indicates the number of STCs used to vote, note that the unit is nanoSTC, with 9 decimal places, the above example is 1 STC.

Note: Replace `your_address` with your own address, do not use all the STCs, leave enough for the gas fee.

3. Abandon Voting

If you want to abort the poll and get back the collateral STCs, you can use the following command.

```
account execute-function -b -s your_address --function 0x1::DaoVoteScripts::revoke_vote -t 0x1::STC::STC -t 0x1::OnChainConfigDao:: OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba --arg 13
```

4. retrieve the collateral STC after the poll

```
account execute-function -b -s your_address --function 0x1::DaoVoteScripts::unstake_vote -t 0x1::STC::STC -t 0x1::OnChainConfigDao:: OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba --arg 13
```

5. View poll results

```
dev call --function 0x1::Dao::proposal_info -t 0x1::STC::STC -t 0x1::OnChainConfigDao::OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba

{
  "ok": [
    13,
    1677045524051,
    1677650324051,
    0,
    0
  ]
}
```

The first number of the return value indicates the proposal id, the second number is the start time, the third number is the end time, the fourth number is the number of for votes and the fifth number is the number of against votes.

**Comment by nkysg** [comment]
## 中文版说明

投票提案需要下载[v1.13.0](https://github.com/starcoinorg/starcoin/releases/tag/v1.13.0)客户端
本提案提议将将链上 Move 字节码版本配置升级到 v6。

## 升级提案信息

- 提案交易: https://stcscan.io/main/transactions/detail/0x786d69ec448b787690244df0adb7051d5990e8db32f06f5c555cad5fb5b7d4f5
- 提案发起人：[0xed9ea1f3533c14e1b52d9ff6475776ba](https://stcscan.io/main/address/0xed9ea1f3533c14e1b52d9ff6475776ba)
- 提案 id: 13

## 升级流程

1.  从今天发起升级提案开始，进行为期 7 天的投票。
2. 7 天后如果投票通过，则进入公示期，公示期 1 天。如果未通过，则提案失败。
3. 公示期过后，任何人都可以发起交易将新的配置写入到链上。该升级交易执行后，Move 新特性激活。 **由于 Move 新特性包含了硬分叉特性，届时尚未升级到 starcoin v1.13.0 版本的节点可能会和主网分叉。**

### 关于 Starcoin 治理的更多信息请参看：

* [Starcoin 技术白皮书中的治理部分](https://developer.starcoin.org/zh/sips/sip-2/)
* [链上治理的生命周期](https://developer.starcoin.org/zh/key_concepts/dao_governance/)
* [通过治理机制修改 DAO 的设置](https://developer.starcoin.org/zh/cli/modify_dao_config/)


注意：投票时，会将当前账户的 STC 抵押到合约中，直到投票结束，当前投票周期为 7 天。期间主动退款则视为放弃投票。** 为了鼓励参与链上治理，未来会推出链上治理的激励措施。本次参与投票的用户会根据激励策略进行补偿。**

## 投票操作指南

### 准备
1. 下载 starcoin 版本 v1.13.0，并同步主网的区块，进入 starcoin 控制台。
2. 将自己的账户导入到节点内置的钱包中，或者从交易所提币到节点钱包的地址中。

### 操作命令

1. 查看投票状态

```
dev call --function 0x1::Dao::proposal_state -t 0x1::STC::STC -t 0x1::OnChainConfigDao::OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba --arg 13
```
第一个参数是提案发起人的地址，第二个参数是提案 id，返回的状态为 1 表示等待状态， 2 表示正在投票中，4 表示投票通过。

2. 执行投票

通过一下命令进行投票：

```
account execute-function -b -s your_address --function 0x1::DaoVoteScripts::cast_vote -t 0x1::STC::STC -t 0x1::OnChainConfigDao::OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba --arg 13 --arg true --arg 1000000000u128
```

第一个参数是提案发起人的地址，第二个参数是提案 id，第三个参数通过 true/false 表示同意和反对，第四个参数表示用来投票的 STC 数量，注意单位是 nanoSTC，有9位是小数点后的单位，上面的例子是 1 STC。

注意：替换 your_address 为你自己的地址，不要用全部的 STC，要留够 gas 费。

3. 放弃投票

如果想放弃投票，取回抵押的 STC，可以用以下命令：

```
account execute-function -b -s your_address --function 0x1::DaoVoteScripts::revoke_vote -t 0x1::STC::STC -t 0x1::OnChainConfigDao::OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba --arg 13
```

4. 投票结束后取回抵押的 STC

```
account execute-function -b -s your_address --function 0x1::DaoVoteScripts::unstake_vote -t 0x1::STC::STC -t 0x1::OnChainConfigDao::OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba --arg 13
```

5. 查看投票结果

```
dev call --function 0x1::Dao::proposal_info -t 0x1::STC::STC -t 0x1::OnChainConfigDao::OnChainConfigUpdate<0x1::LanguageVersion::LanguageVersion> --arg 0xed9ea1f3533c14e1b52d9ff6475776ba

{
  "ok": [
    13,
    1677045524051,
    1677650324051,
    0,
    0
  ]
}
```

返回值的第一个数表示提案 id，第二个数是开始时间，第三个数是结束时间，第四个数是投票同意的票数，第五个数是投票反对的票数。

**Comment by Garasptr** [comment]
Let's gooo !!!

0x932575681e1b26da0cab8cf854440537
